include/deal.II/lac/linear_operator.h

```diff
@@ -0,0 +1,1175 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2010 - 2015 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef __deal2__linear_operator_h
+#define __deal2__linear_operator_h
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/exceptions.h>
+#include <deal.II/lac/vector_memory.h>
+
+#ifdef DEAL_II_WITH_CXX11
+
+#include <functional>
+#include <array>
+
+DEAL_II_NAMESPACE_OPEN
+
+// Forward declarations:
+
+template <typename Number> class Vector;
+template <typename Number> class BlockVector;
+
+template <typename Range, typename Domain> class LinearOperator;
+
+template <typename Range = Vector<double>,
+          typename Domain = Range,
+          typename Exemplar,
+          typename Matrix>
+LinearOperator<Range, Domain> linop (const Exemplar &, const Matrix &);
```

Comment:
My preference would be no namespace and long names `LinearOperator`, `linear_operator`, `transpose_operator`, `inverse_operator`, `identity_operator`, `block_operator`, `block_diagonal_operator`.

You could also put all of these into a namespace of their own and then `using` it in the `dealii` namespace.

Out of curiosity, where (except in `inverse_operator`) do you actually make use of the fact that the operator is *linear*? It seems to me that that is the *intention* for all of these objects, but not necessarily a requirement for your algorithms.